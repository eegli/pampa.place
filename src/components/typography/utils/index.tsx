import Box from '@mui/material/Box';

export const em = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
