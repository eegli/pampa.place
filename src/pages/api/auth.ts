// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import type { NextApiRequest, NextApiResponse } from 'next';
import { env } from 'process';

export type AuthReq = {
  pw: string;
};

export type AuthRes = {
  apikey: string;
};

const accessPW = env.APP_ACCESS_PW || '';

function getApiKeyFromEnv() {
  switch (env.NODE_ENV) {
    case 'development':
      return env.MAPS_LOCAL_API_KEY || '';
    case 'production':
      return env.MAPS_PROD_API_KEY || '';
    default:
      return '';
  }
}

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<AuthRes>
) {
  const envApiKey = getApiKeyFromEnv();
  if (!envApiKey) {
    throw new Error('Missing API keys');
  }

  if (!accessPW) {
    throw new Error('No password defined');
  }

  if (!req.query.pw) {
    res.status(400).end('Resource not found');
  }
  if (req.query.pw === accessPW) {
    res.status(200).json({ apikey: envApiKey });
  } else {
    res.status(401).end('Invalid password');
  }
}
