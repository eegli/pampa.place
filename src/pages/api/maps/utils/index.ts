export {apiData as data} from '@/config/maps';
import {NextApiHandler} from 'next';

type ErrorResponse = {
  error: string;
};

type DataResponse<T> = T extends unknown[]
  ? {
      data: T;
    }
  : {
      data: T[];
    };

export type ApiJSONHandler<T = unknown> = NextApiHandler<
  DataResponse<T> | ErrorResponse
>;
export type ApiTextHandler = NextApiHandler<string> | ErrorResponse;

export function nthQuery<T>(arg: T | T[], index = 0): T | undefined {
  return Array.isArray(arg) ? arg[index] : arg;
}
