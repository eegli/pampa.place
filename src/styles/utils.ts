import {keyframes} from '@mui/material/styles';

type KeyedColorFadeProps = [number, string][];

export const keyedColorFade =
  (prop: string) =>
  (...steps: KeyedColorFadeProps) => {
    const gradient = steps
      .map(([perc, color]) => `${perc}% {${prop}: ${color}}`)
      .join(' ');
    return keyframes(gradient);
  };
