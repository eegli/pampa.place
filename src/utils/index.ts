// Area in square kilometers, distance in meters
import { LatLngLiteral } from '@config';
import * as turf from '@turf/turf';

// https://www.desmos.com/calculator/xlzbhq4xm0
export function calculateScore(area: number, dist: number) {
  if (dist < 0) return 0;
  const c =
    Math.log(area + 1) * Math.sqrt(dist * area) + Math.log(dist ** 2 + 1);
  const score = 5000 * Math.E ** -(dist / c);

  return Math.round(score);
}
// Returns distance between two points in kilometers

export function calculateDistance(
  p1: LatLngLiteral,
  p2: LatLngLiteral,
  units: turf.Units = 'kilometers'
) {
  const from = turf.point([p1.lng, p1.lat]);
  const to = turf.point([p2.lng, p2.lat]);
  return turf.distance(from, to, { units });
}
// Formats distance
export function formatDist(km: number, toFixed = 1) {
  if (km < 0) return '-';
  if (km > 1000) {
    return (km / 1000).toFixed(toFixed) + ' m';
  }
  return km.toFixed(toFixed) + ' km';
}

export function max(num: number, limit: number) {
  return num >= limit ? limit : num;
}

export function id(): string {
  const array = new Uint32Array(8);
  window.crypto.getRandomValues(array);
  let str = '';
  for (let i = 0; i < array.length; i++) {
    str += (i < 2 || i > 5 ? '' : '-') + array[i].toString(16).slice(-4);
  }
  return str;
}
