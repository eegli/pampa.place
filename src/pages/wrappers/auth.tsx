import {Status, Wrapper} from '@googlemaps/react-wrapper';
import {ReactNode, useEffect} from 'react';
import {Spinner} from '../../components/feedback/spinner';
import {useAppSelector} from '../../redux/hooks';
import {PageContentWrapper} from '../../styles/containers';
import {Login} from './login';

export const AuthWrapper = ({children}: {children?: ReactNode}) => {
  const apiKey = useAppSelector(s => s.app.apiKey);

  // https://developers.google.com/maps/documentation/javascript/events
  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (window as any).gm_authFailure = () => {
      // Maybe a TODO
    };
  });

  // An empty string as api key is allowed for development mode
  if (typeof apiKey !== 'string') return <Login />;

  return (
    <Wrapper
      apiKey={apiKey}
      version="3.47.2"
      render={status => {
        switch (status) {
          case Status.SUCCESS:
            return <>{children}</>;
          case Status.LOADING:
            return (
              <PageContentWrapper id="spinner">
                <Spinner />
              </PageContentWrapper>
            );
          default:
            return (
              <PageContentWrapper id="spinner">
                <div>mega fail</div>
              </PageContentWrapper>
            );
        }
      }}
    />
  );
};
