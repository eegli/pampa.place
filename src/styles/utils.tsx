import Box from '@mui/material/Box';
import {keyframes} from '@mui/material/styles';

type KeyedColorFadeProps = [number, string][];

export const keyedColorFade =
  (prop: string) =>
  (...steps: KeyedColorFadeProps) => {
    const gradient = steps
      .map(([perc, color]) => `${perc}% {${prop}: ${color}}`)
      .join(' ');
    return keyframes(gradient);
  };

export const em = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
