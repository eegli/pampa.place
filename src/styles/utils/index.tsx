import createCache from '@emotion/cache';
import Box from '@mui/material/Box';
import {keyframes} from '@mui/material/styles';

const isBrowser = typeof document !== 'undefined';

// On the client side, Create a meta tag at the top of the <head> and set it as insertionPoint.
// This assures that MUI styles are loaded first.
// It allows developers to easily override MUI styles with other styling solutions, like CSS modules.
export function createEmotionCache() {
  let insertionPoint;

  if (isBrowser) {
    const emotionInsertionPoint = document.querySelector<HTMLMetaElement>(
      'meta[name="emotion-insertion-point"]'
    );
    insertionPoint = emotionInsertionPoint ?? undefined;
  }

  return createCache({key: 'mui-style', insertionPoint});
}

export const keyedColorFade =
  (prop: string) =>
  (...steps: [number, string][]) => {
    const gradient = steps
      .map(([perc, color]) => `${perc}% {${prop}: ${color}}`)
      .join(' ');
    return keyframes(gradient);
  };

export const colorize = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
