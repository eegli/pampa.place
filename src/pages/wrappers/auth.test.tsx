import {createMockState, createMockStore, render} from '@/tests/utils';
import * as GoogleMapsReactWrapper from '@googlemaps/react-wrapper';
import {AuthWrapper} from './auth';

const wrapperSpy = jest
  .spyOn(GoogleMapsReactWrapper, 'Wrapper')
  .mockImplementation(({children}) => {
    return <div id="rendered">{children}</div>;
  });

beforeEach(() => {
  wrapperSpy.mockClear();
});

describe('Auth wrapper', () => {
  ['apikey', '', undefined].forEach(apiKey => {
    it(`inits Google Maps with api key value: ${apiKey}`, () => {
      const state = createMockState({
        app: {
          apiKey,
        },
      });
      const store = createMockStore(state);
      render(<AuthWrapper />, store);
      // Valid api keys
      if (typeof apiKey === 'string') {
        expect(wrapperSpy).toHaveBeenCalledTimes(1);
        expect(wrapperSpy.mock.calls[0][0]).toMatchSnapshot(apiKey);
        expect(wrapperSpy.mock.results[0].value).toMatchSnapshot(apiKey);
      } else {
        // undefined is not a valid key
        expect(wrapperSpy).not.toHaveBeenCalled();
      }
    });
  });
});
