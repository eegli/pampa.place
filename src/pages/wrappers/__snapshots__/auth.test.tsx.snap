// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth wrapper inits Google Maps with api key value:  1`] = `
Object {
  "apiKey": "",
  "render": [Function],
  "version": "3.47.2",
}
`;

exports[`Auth wrapper inits Google Maps with api key value:  2`] = `
<div
  id="rendered"
/>
`;

exports[`Auth wrapper inits Google Maps with api key value: apikey: apikey 1`] = `
Object {
  "apiKey": "apikey",
  "render": [Function],
  "version": "3.47.2",
}
`;

exports[`Auth wrapper inits Google Maps with api key value: apikey: apikey 2`] = `
<div
  id="rendered"
/>
`;
